mlir/include/mlir/IR/OpImplementation.h

```diff
@@ -1715,7 +1715,10 @@ class OpAsmDialectInterface
     OverridableAlias,
     /// An alias was provided and it should be used
     /// (no other hooks will be checked).
-    FinalAlias
+    FinalAlias,
+    /// A dialect alias was provided and it will be used
+    /// (no other hooks will be checked).
+    DialectAlias
```

Comment:
`DialectAlias` is different from the `TypeAsmAliasTypeInterface`.
- With `DialectAlias` you can alias any type or attribute, however, the type might not always be printed using the alias.
- `TypeAsmAliasTypeInterface`, you can only alias objects that allow aliases. So it has more constraints in what it can do.

For example, with the first mechanism you can alias `IntegerType::get(2^exp)` to `p2_int<exp>`, and keep `i3` as `i3`, however the second mechanism allows aliasing the type to only one dialect (unless the type implements a specific alias rule).

Reply:
Is there actually a use case for this and do you need that powerful of a mechanism for the pointer dialect? In the interest of landing this PR with the interface mechanism I'd just drop that mechanism and implement it in another PR of really needed.

Reply:
Not really. I was just trying to solve the problem in general, and I'm open drop it. However, I think there's no rush with landing this, as Euro LLVM will happen soon, hence all the other Ptr patches are not likely to land soon.

I'll implement the DenseMap approach to talk with real numbers. If you're interested, we can also talk more about it in the meeting.